Feature/2021 05 25 intro to mqtt (#666)
ordina-jworks/ordina-jworks.github.io: _posts/2021-05-26-intro-to-mqtt.md

- Time decoupling: Publisher and subscribers do not need to run at the same point in time.
- Synchronization decoupling: Operations do not need to be interrupted during publishing or receiving.

Also be reminded that MQTT is **NOT** a message queue!
By default messages will not be stored if there are no clients to consume them, and even retained messages only keep the last retained message, overwriting any previous retained message.

<div style="text-align: center; margin: 0px auto;">
    <a href="{{ '/img/2021-05-26-intro-to-mqtt/mqtt-publish-subscribe.png' | prepend: site.baseurl }}" data-lightbox="ui" data-title="MQTT">
        <img alt="MQTT" src="{{ '/img/2021-05-26-intro-to-mqtt/mqtt-publish-subscribe.png' | prepend: site.baseurl }}" class="image fit" style="margin:0px auto; width: 100%; display: inline-block;">
    </a>
</div>

## MQTT terminology

To better understand and grasp the concepts of the MQTT protocol explained in this blog post I have listed some terminology below:

[Broker](https://www.hivemq.com/blog/mqtt-essentials-part-3-client-broker-connection-establishment/){:target="_blank" rel="noopener noreferrer"}:
The host that acts as the manager for all messages.

> It is responsible for receiving messages from publishers, performing checks to see which subscribers match the topic and sending the messages to said subscribers.

Reviewer: said subscribers? shouldn't it be those subscribers?

[Client](https://www.hivemq.com/blog/mqtt-essentials-part-3-client-broker-connection-establishment/){:target="_blank" rel="noopener noreferrer"}:
An MQTT client, meaning any instance that implements logic to connect to a broker.

> Both publishers and subscribers are clients.

Reviewer: Clients can be both publisher and subscriber

Publishers send messages and subscribers consume messages.

> A client can implement both sides and act as both a publisher and a subscriber.

Reviewer: Needs some rephrasing with the above

[Last Will and Testament (LWT)](https://www.hivemq.com/blog/mqtt-essentials-part-9-last-will-and-testament/){:target="_blank" rel="noopener noreferrer"}:
Is a feature to notify clients about a client that has disconnected in an ungraceful manner.
The message is sent to the broker when a client connects so it can be sent to other clients later on if required.
If the client disconnects gracefully the broker discards the LWT message.

## Broker options

As MQTT requires a broker instance to function choosing the right one is crucial.
There many different options available, both run run locally or hosted in the cloud.

There are also different versions of the MQTT protocol, not every broker supports all the different versions:

- 3.1: The older IBM based version, less used these days
- 3.1.1: OASIS standard compliant, the most used version nowadays
- 5: The newest version (2019), not yet widely used

### [Eclipse Mosquitto](https://mosquitto.org/){:target="_blank" rel="noopener noreferrer"}

Eclipse Mosquitto is an open source implementation of an MQTT message broker.

> It supports all three mayor versions of the protocol.

Reviewer: ```suggestion
It supports all three major versions of the protocol.
```

Reviewer: remove this one and merge with below

> The broker is very lightweight and can run on low powered devices like the Raspberry Pi.

Reviewer: ```suggestion
The broker supports all three major versions and can run on low-powered devices like the Raspberry Pi.
```

> The broker is very lightweight and can run on low powered devices like the Raspberry Pi.

Reviewer: assuming the protocol is lightweight, the broker should also be lightweight

I use this one at home for my home automation projects.

> Installing it is very easy (rbpi on debian buster):

Reviewer: Rephrase this, (It is very easy to install on a Debian buster)

```shell
wget http://repo.mosquitto.org/debian/mosquitto-repo.gpg.key
sudo apt-key add mosquitto-repo.gpg.key
cd /etc/apt/sources.list.d/

> # Pick the correct url for your flavour of debian (we pick buster as the default):

Reviewer: ```suggestion
# Pick the correct URL for your flavour of Debian (we pick buster as the default):
```

sudo nano /etc/mosquitto/mosquitto.conf
# At the end of the file add:
# allow_anonymous false
# password_file /etc/mosquitto/credentials
sudo service mosquitto restart
```

### [Aedes](https://github.com/moscajs/aedes){:target="_blank" rel="noopener noreferrer"}

Aedes is the follow up/split from Mosca and is fully open source.
It is a node based MQTT broker that is scalable and lightweight.
The broker only has support for the 3.1 and 3.1.1 protocol versions, 5.0 is not supported yet.

Installing is very simple, just make sure you have node installed and simply install it by using npm: `npm install aedes`.
You are responsible for creating the server instance from code.
A very basic implementation of the broker is:

```js
const aedes = require('aedes')();
const server = require('net').createServer(aedes.handle);
const port = 1883;

server.listen(port, () => {
  console.log('Server started and listening on port ', port);
});
```

### [HiveMQ](https://www.hivemq.com/){:target="_blank" rel="noopener noreferrer"}

HiveMQ is an MQTT-based platform that includes a broker.
It has the option to be hosted in the cloud (with a free trial tier) or to be run locally.

> The broker has support for all three mayor versions of the protocol.

Reviewer: ```suggestion
The broker has support for all three major versions of the protocol.
```

It does require you to create an account before you can use the cloud tier or even download the zip package for local installation.

### [Emqttd](https://emqtt.io/docs/v1/index.html){:target="_blank" rel="noopener noreferrer"}

Emqttd is another fully open source broker.
The project is written in Erlang and is fully compatible with the 3.1 and 3.1.1 versions of the protocol.

### [VerneMQ](https://vernemq.com/){:target="_blank" rel="noopener noreferrer"}

> VerneMQ is another well known broker that is also fully open source, and written in Erlang.

Reviewer: ```suggestion
VerneMQ is another well-known broker that is also fully open-source and written in Erlang.
```

It has the ability to scale very well, both vertically and horizontally.

> The broker has support for all three mayor versions of the protocol.

Reviewer: ```suggestion
The broker has support for all three major versions of the protocol.
```

> In addition to the free to use broker they also have paid tiers of support.

Reviewer: ```suggestion
In addition to the free-to-use broker, they also have paid tiers of support.
```

## Basic examples

Basic C example (for use on an ESP-01):

```cpp
#include "EspMQTTClient.h"

EspMQTTClient *client;

void setup()
{
  client = new EspMQTTClient("SSID", "SSID_PASS", "BROKER_URL", "MQTT_USERNAME", "MQTT_PASSWORD", "CLIENT_NAME", 1883);
  //client->enableDebuggingMessages(true);

  client->enableLastWillMessage("LAST_WILL_TOPIC", "LAST_WILL_MESSAGE");
}

void onConnectionEstablished()
{
    Serial.println("MQTT: Connected");
    client->publish("TOPIC", "connected");

    client.subscribe(actionTopic, [](const String & payload) {
        Serial.println(payload);
    });
}

void loop()
{
  client->publish("TOPIC", "DATA");
  sleep(1000);
}

> 

Reviewer 2: Remove empty line

    constructor() {
        this.client = this.connectMqtt();
        this.receive();
    }

    private connectMqtt(): any {
        const client: Client = mqtt.connect('mqtt://broker-url:1883', {username: 'username', password: 'password'});

        client.on('connect', () => {
            console.log('Connected to MQTT broker!');
        });

        return client;
    }

    private receive(): void {
        this.client.subscribe('topic', {qos: 0}, (err: Error, granted: ISubscriptionGrant[]) => {
            console.log(granted);

            if (granted && granted.length === 1) {
               this.client.on('message', (topic: string, payload: Buffer, packet: Packet) => {
                    console.log(payload.toString());
               });
            }
        })
    }
}

const main = new Main();
```

The node code is a bit more complex but allows you to create more complex applications.
It uses the [MQTT.js](https://www.npmjs.com/package/mqtt){:target="_blank" rel="noopener noreferrer"} library which has very good and extensive documentation.
This is not meant to run on the IoT device but on a separate device that reacts to messages from the IoT devices.

> Please not that this will not manage the WiFi/ethernet connection of the device that it is running on, that is left to the OS/User.

Reviewer 2: Note
.. running on. This is left to..

> Please not that this will not manage the WiFi/ethernet connection of the device that it is running on, that is left to the OS/User.

Reviewer: ```suggestion
Please note that this will not manage the WiFi/ethernet connection of the device that it is running on, which is left to the OS/User.
```

## Conclusion

MQTT is an ideal protocol to use for lightweight communication on ip enabled devices.
The pub/sub architecture allows for a decoupled environment of clients that can operate independently of each other.

> Thanks to the protocol and its implementations being very lightweight it is very handy to use in combination with IoT and home automation projects.

Reviewer 2: Lightweight, it

Getting started with some simple setup is relatively easy and does not require difficult programming to wire things up.

> In a future blog post I will go a bit more in depth and show off a project which makes use of MQTT to wire devices together.

Reviewer 2: blog post, I will
